v1/image.go

```diff
@@ -55,4 +58,29 @@ type Image interface {
 	// LayerByDiffID is an analog to LayerByDigest, looking up by "diff id"
 	// (the uncompressed hash).
 	LayerByDiffID(Hash) (Layer, error)
+
+	//////////////////////////////////////////////////////
+	// All method below are for setting functional options
+	//////////////////////////////////////////////////////
+
+	// SetTransport allows a transport configuration to be passed down
+	// to the underlying implementation in the image.
+	SetTransport(http.RoundTripper) error
+
+	// SetAuth allows auth configuration to be passed down to the
+	// underlying http.Client.
+	SetAuth(authn.Keychain) error
+
+	// SetOpener sets the opener for an image in the local daemon,
+	// buffering or not based on the provided boolean.
+	SetOpener(bool) error
```

Comment:
tl;dr I'm uncomfortable putting these here.

I feel like having these here leaks implementation details and forces all implementations to surface details that are only useful to a few.

I also don't want us to end up with truly mutable images.  I like the functional construction that yields frozen objects, and I don't like the prospect that someone can call `SetAuth` in the middle of one of my interactions with an `Image`.  Now everything needs synchronization, yuck :(

I think the design I'd go with is something like:
```golang
package v1

// Implementations extend Image with the core logic.
type Image interface {
  ...
}

// Implementations extend ImageOpener with options.
type ImageOpener interface {
  // Open returns a frozen Image object that can no longer be configured.
  Open() (Image, error)
}

// Implementations would define:
// type Option func(ImageOpener) error
```

In this world, `remote.Image` would change something like:
```golang
package remote

// public for Option signatures.
type ImageOpener struct {
  // package private fields
}
func NewImageOpener(name.Reference) (ImageOpener, error) { ... }

// Signature for our options.
type Option func(ImageOpener) error

// Option constructors (can access package private fields)
func WithAuth(auth.Authenticator) Option { ... }
func WithTransport(http.RoundTripper) Option { ... }

func Image(ref name.Reference, opts ...Option) (v1.Image, error) {
  o, err := Opener(ref)
  for _, opt := range opts {
    err := opt(o)
  }
  return o.Open()
}
```

WDYT?

Reply:
Great point about the sync stuff, hadn't thought about that. The reason I created the concept of the `MutableImage` in the first place was just so we could have something to operate on while we apply the options, and then return its embedded `v1.Image` when we're done, but where I went wrong was not making all of the option methods on the `MutableImage`.

Either way, I like your design better, let me play around with it a bit and see how it looks.